### PR TITLE
fix: logger initialization before flags parsing

### LINKED ISSUE #7124: bug(log): incorrect log format for flag messages
## Description
Incorrect log format for flag messages:
```
2024/07/09 10:08:31 WARN '--vuln-type' is deprecated. Use '--pkg-types' instead.
2024-07-09T10:08:31+04:00	INFO	Vulnerability scanning is enabled
```

We need to init logger before these messages.

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
We need to init logger before flags parsing to avoid wrong log messages format.

before:
```bash
➜  trivy -d image alpine -c config.yaml --vuln-type os 
2024/08/22 14:09:16 WARN '--vuln-type' is deprecated. Use '--pkg-types' instead.
2024/08/22 14:09:16 INFO Loaded file_path=config.yaml
2024-08-22T14:09:16+06:00       DEBUG   Parsed severities       severities=[UNKNOWN LOW MEDIUM HIGH CRITICAL]
...

➜ trivy -q image alpine -c config.yaml --vuln-type os  
2024/08/22 16:22:17 WARN '--vuln-type' is deprecated. Use '--pkg-types' instead.
2024/08/22 16:22:17 INFO Loaded file_path=config.yaml

alpine (alpine 3.20.2)

Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)

```

after
```bash
➜  trivy -d image alpine -c config.yaml --vuln-type os
2024-08-22T14:08:58+06:00       WARN    '--vuln-type' is deprecated. Use '--pkg-types' instead.
2024-08-22T14:08:58+06:00       INFO    Loaded  file_path="config.yaml"
2024-08-22T14:08:58+06:00       DEBUG   Parsed severities       severities=[UNKNOWN LOW MEDIUM HIGH CRITICAL]
...

➜ trivy -q image alpine -c config.yaml --vuln-type os

alpine (alpine 3.20.2)

Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)
```

## Related issues
- Close #7124

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [ ] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/commands/app.go [hunk 1]
```diff
@@ -65,6 +65,12 @@ Use "{{.CommandPath}} [command] --help" for more information about a command.{{e
 // NewApp is the factory method to return Trivy CLI
 func NewApp() *cobra.Command {
 	cobra.EnableTraverseRunHooks = true // To execute persistent pre-run hooks from all parents.
+
+	// Initialize the logger for `flag` and `app` packages.
+	// To display logs from these packages in Trivy format.
+	// We will set the `debug` and `disable` format after parsing the `--debug` and `--quiet` flags.
+	log.InitLogger(false, false)
```
Comment (DmitriyLewen) [inline pkg/commands/app.go:72]: I simply added init logger before parsing flags.
But there is a problem with this way:
if the `--quiet` flag is used - we still show the log for logs from "app" and "flags".

As another way - I added `DeferredLogger` to save log messages and print them on command (in our case after logger initialization)
See

File: pkg/log/deferred.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+func (d *deferredLogger) Error(message string, args ...any) {
+	d.deferredLogs = append(d.deferredLogs, deferredLog{
+		logFunc: Error,
+		message: message,
+		args:    args,
+	})
+}
+
+func (d *deferredLogger) PrintLogs() {
```
Comment (DmitriyLewen) [inline pkg/log/deferred.go:50]: We can add this into:
https://github.com/aquasecurity/trivy/blob/6ad109849085aab8cfb1ae04aa9a4b4315d59911/pkg/log/logger.go#L38-L42

But perhaps it won't be entirely obvious

**Review by knqyf263 — COMMENTED** [review]
I like the idea of DeferredLogger, but the current implementation allows us to miss log emissions before logger initialization. For example, although `LoadAll()` is called before parsing CLI flags, we forgot to replace it with `log.DeferredLogger`.
https://github.com/aquasecurity/trivy/blob/88ba46047c93e6046292523ae701de774dfdc4dc/pkg/plugin/manager.go#L281

Thus, I think we should replace the default logger or handler.

Also, we need to handle `WithPrefix` and `WithAttrs`, but the current implementation doesn't work with them.
https://github.com/aquasecurity/trivy/blob/88ba46047c93e6046292523ae701de774dfdc4dc/pkg/plugin/manager.go#L65

To resolve these problems, I implemented `DeferredHandler`. PTAL, @DmitriyLewen.

**Comment by DmitriyLewen** [comment]
Nice upgrade of my idea!
Looks good.

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**